### PR TITLE
Add the CLI version to telemetry events

### PR DESCRIPTION
**Pull request by aeisenberg**
This adds the CLI version to telemetry command-usage events.

Note that the CLI server is created after the telemetry listener is created. The first few telemetry events may have a "not-set" value for the CLI version.

## Checklist

- [n/a] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [n/a] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

File: extensions/ql-vscode/src/telemetry.ts [hunk 1]
```diff
@@ -157,12 +160,18 @@ export class TelemetryListener extends ConfigListener {
       ? CommandCompletion.Cancelled
       : CommandCompletion.Failed;
 
+    const cliVersion = this._cli
+      ? (await this._cli.getVersion()).toString()
```
Comment (koesie10) [inline extensions/ql-vscode/src/telemetry.ts:164]: Should we be catching the error here and setting the `cliVersion` to something like `errored`? It seems like we could start missing command usages if the CLI server is in a bad state.

**Review by koesie10 — COMMENTED** [review]
Thanks for adding this, this will definitely be useful!

I'm somewhat concerned about the performance impact this will have on sending commands. If there is no version available, this will spawn a process to get the CLI version, which might delay the telemetry and possibly skip sending the telemetry if VS Code gets closed early.

I also think the CLI version would be useful on other telemetry events (especially errors). I think it would be good if we could use a `addTelemetryProcessor` to add the CLI version tag to every telemetry event we send. The telemetry processor is not `async`, but we could use the cached CLI version here. That would also solve my concern about performance. We should already always be retrieving the version in the CLI server (since it's also used by CLI version constraints and the status bar), so I think this would be a very viable option for adding the CLI version to all telemetry events without the possibility of spawning a process on every telemetry event.

**Comment by robertbrignull** [comment]
I agree with Koen's points. Calculating the CLI version while handling the telemetry event doesn't seem great, both in terms of performance and the chance of errors happening and stopping us sending telemetry. Could we set the CLI version once, and then maybe update it if the CLI changes? I also agree it would be nice to have this on all events, like how we include `isCanary`.

File: extensions/ql-vscode/src/telemetry.ts (hunk 1 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:164]: After the change, this should no longer be relevant.

**Comment by aeisenberg** [comment]
I think I addressed all the comments.

**Comment by aeisenberg** [comment]
Not sure why I'm getting a failing test:

https://github.com/github/vscode-codeql/actions/runs/4693003381/jobs/8320617631#step:9:69

It doesn't seem related to my changes, but it is repeatable in CI.

**Comment by aeisenberg** [comment]
Maybe updating the branch will help.

**Comment by aeisenberg** [comment]
Trying to run locally, but getting this error on startup:

> [GitHub.vscode-codeql]: Menu item references a command `codeQL.debug.quickEvalContextEditor` which is not defined in the 'commands' section.

**Comment by aeisenberg** [comment]
Fixed by adding this to package.json.
```
      {
        "command": "codeQL.debug.quickEval",
        "title": "CodeQL: Debug Quick Evaluation"
      },
      {
        "command": "codeQL.debug.quickEvalContextEditor",
        "title": "CodeQL: Debug Quick Evaluation"
      },
```

But these commands are not yet hooked up to any handlers.

**Comment by aeisenberg** [comment]
Nope.

**Comment by robertbrignull** [comment]
I'd guess we have some tests that are very close to the timeout value, and they've just tipped over recently 😢

**Comment by robertbrignull** [comment]
We have an internal issue about tests timing out and I've mentioned this there. Either we should look at making the tests faster or raising the timeout value.

**Review by koesie10 — APPROVED** [review]
LGTM, thanks for addressing our concerns!

**Review by robertbrignull — APPROVED** [review]
LGTM too

**Comment by aeisenberg** [comment]
Here's my analysis:

[The first error](https://github.com/github/vscode-codeql/actions/runs/4693944558/jobs/8321559046?pr=2307#step:9:129) was indeed a timeout error: 

```
    thrown: "Exceeded timeout of 20000 ms for a test.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."
```

The tests re-ran and somewhere during the retry, the original query finally completed, but the callback failed since the connection had already been disposed. So, I think increasing the timeout will work, and we should also figure out a way to properly cancel a query in the teardown.

Also, I found another error where `removeAllDatabases` throws. This method is called during teardown and makes a call to the query server to de-register the database. If the query server is already shut down, then an error will be thrown. But if the query server is shut down, then deregistering isn't necessary, so it's ok to ignore. 

I will put a fix in for this and bump the timeouts. Both will be in a separate PR.

**Pull request merged**